src/object/cli_obj.c

```diff
@@ -5062,30 +5065,58 @@ dc_obj_query_key(tse_task_t *api_task)
 	D_ASSERT(!obj_auxi->args_initialized);
 	D_ASSERT(d_list_empty(head));
 
-	/* In each redundancy group, the QUERY RPC only needs to be sent
-	 * to one replica: i += replicas
-	 */
-	for (i = 0; i < obj->cob_shards_nr; i += replicas) {
+	step = replicas;
+	is_ec_obj = obj_is_ec(obj);
+	for (i = 0; i < obj->cob_shards_nr; i += step) {
 		tse_task_t			*task;
 		struct shard_query_key_args	*args;
+		struct daos_oclass_attr		*oca;
 		int				 shard;
 
 		if (api_args->flags & DAOS_GET_DKEY) {
+			/* for EC obj, will query from all data shards if
+			 * obj_grp_leader_get failed (no parity available).
+			 */
+			if (ec_query_data_tgt) {
+				shard = i;
+				query_nr++;
+				D_ASSERT(query_nr <= tgt_nr);
+				if (query_nr == tgt_nr) {
+					i -= (tgt_nr - 1);
+					D_ASSERT(i >= 0 && (i % replicas == 0));
+					step = replicas;
+					ec_query_data_tgt = false;
+				}
+				goto shard_task_create;
+			}
 			/* Send to leader replica directly for reducing
 			 * retry because some potential 'prepared' DTX.
 			 */
 			shard = obj_grp_leader_get(obj, i, map_ver);
 			if (shard < 0) {
-				rc = shard;
-				D_ERROR(DF_OID" no valid shard, rc "DF_RC".\n",
-					DP_OID(obj->cob_md.omd_id),
-					DP_RC(rc));
-				D_GOTO(out_task, rc);
+				if (!is_ec_obj) {
+					rc = shard;
+					D_ERROR(DF_OID" no valid shard, rc "
+						DF_RC".\n",
+						DP_OID(obj->cob_md.omd_id),
+						DP_RC(rc));
+					D_GOTO(out_task, rc);
+				}
+				oca = obj_get_oca(obj, true);
```

Comment:
obj_get_oca() can be called only once and kept when used next time.

Reply:
yes, I want the client-side only do one time oca search and keep/reuse it on obj (I'll refine some code later), and server-side at most do it one time (probably need not do it for some cases).